man/iperender.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics, 
.\" respectively.
.TH IPERENDER 1 "June 27, 2009"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
iperender \- Export Ipe document as bitmap or SVG 
.SH SYNOPSIS
.B iperender
( -svg | -png ) 
[ -page \fIpage\fP ] 
[ -view \fIview\fP ]
[ -resolution \fIdpi\fP ]
[ -transparent ] 
\fIinput-file\fP \fIoutput-file\fP

.SH DESCRIPTION
.PP
\fBiperender\fP exports an Ipe document as a PNG bitmap or in SVG format.
Options are:
.TP
\fB-png\fP
convert to PNG format
.TP
\fB-svg\fP
convert to SVG format
.TP
\fB-page\fP \fIpage\fP
export this page from a multipage document.
.TP
\fB-view\fP \fIview\fP
export this view from a page with multiple views.
.TP
\fB-resolution\fP \fIdpi\fP
set resolution for exporting to PNG.
.TP
\fB-transparent\fP
make background transparent when exporting to PNG.

.SH ENVIRONMENT VARIABLES

\fBipetoipe\fP respects the following environment variables:

.TP
\fBIPELATEXDIR\fP
the directory where \fBipetoipe\fP runs Pdflatex.
.TP
\fBIPEFONTMAP\fP
the complete path of the font map, describing where Ipe can find the
14 standard PDF fonts.
.TP
\fBIPEDEBUG\fP
set this to 1 for debugging output from \fBiperender\fP.

.SH AUTHOR
Otfried Cheong

.SH REPORTING BUGS
.ad l
Please report bugs using Ipe bugzilla at
.I "http://ipe7.sourceforge.net/bugzilla.html"

.SH SEE ALSO
.ad l
Ipe is documented fully in
.IR "The Ipe Manual" ,
which is available online at
.I "http://ipe7.sourceforge.net/manual/manual.html" .


news.txt
--------------------------------------------------------------------

Ipe 7.0.11 (2010-11-13) SVN 444 India edition

 * Fixed definition of filled marks in basic.isy and presentation.isy.

 * Fixed loading of ipelets to occur before analyzing the
   configuration, to make customization work as documented.

 * "version" attribute in <ipe> tag was incorrectly filled with the
   Ipelib version.  Now using the file format version as intended.

 * Stylesheet can specify an alternate encoding for the Latex source
   file.  The Latex source is converted from UTF-8 to this encoding
   during Latex conversion.

 * Possibility to store notes with each page.

 * Fixed selection status in "Insert image" ipelet.

 * Fixed bug in Ipe6 stylesheet in ipe6upgrade.

 * Fixed bugs in alignment ipelet (thanks Norbert).

 * Added "mactex2009-fontmap.xml" (by Changmook Chun).

 * Use -dylib_install_name option on Mac OS.

 * Include "beamer.isy" to use math fonts as in beamer.

 * Changed temporary latex filename to "ipetemp.tex"

 * Small changes to make Ipe compile with Sun compiler (thanks
   Kostas).

 * Fixed error reading views with no visible layer.

 * New preference prefs.automatic_use_title.

 * New -transparent flag for Iperender.

 * Possibility to clip to paper in ipeqt::Canvas.

 * config.screen_geometry provides width and height of default screen
   in Lua.

 * New preference prefs.auto_external_editor.

 * Shortcuts in linetool can be configured (#346).

 * Fixed display of line cap and line join in EPS (bug #331).

 * Ipe should now work with Texlive 2010 (bug #350).

 * Fixed bug in undo clipping (bug #333).

 * Lua code can show and hide dock windows using showTool.

--------------------------------------------------------------------

Ipe 7.0.10 (2009-12-09) SVN 402

 * Fixed crash in arc drawing (bug #324).
 
 * Allow customization of canvas display: paper color, classic grid,
   thickness of grid lines (see prefs.lua).

 * Fixed display in path drawing tool on Windows (bug #323).

 * Fixed broken line cap on circular arc (bug #325).

 * Preference to set size of main window at startup
   (prefs.window_size).

 * SONAME set for all dynamic libraries, and symbolic links created
   while building and installing.

 * The "change width" box was placed incorrectly if the minipage had
   been moved (bug #326).

--------------------------------------------------------------------

Ipe 7.0.9 (2009-11-20) SVN 393

 * Changed default setting for MOC to moc-qt4.

 * IPEANCIENTPDFTEX environment variable to support pdftex versions
   older than 1.40.

 * Support for color-keyed bitmaps.

 * Fixed snapping to intersections of ellipses (bug #318).

 * Convert "runlatex.bat" to OEM codepage on Windows.

 * Open/Save can now handle non-ASCII filenames when not using UTF-8.

 * New script "ipe-update-master" (see manual under "advanced topics").

 * Better support for Mac OS 10.6 compilation.

 * Support non-ASCII characters in PDF output of document
   title/subject/author and in bookmarks.

 * config.platform now returns "apple" on Mac OS X.

 * Fixed snapping to boundaries of arcs.

 * Implemented F3 (Set line & snap) for arcs and Bezier splines.

--------------------------------------------------------------------

Ipe 7.0.8 (2009-11-04) SVN 373

 * Option -show-configuration to show the current configuration on
   stdout. 

 * Recognize Mac OS and change some compilation flags in Makefiles.
 
 * FreeBSD and Mac OS do not support "install -t <dir>", and so
   invocations of "install" were changed to use the traditional syntax
   "install <files> <dir>".

 * Fixed wrong transformation matrix after path editing (bug #313).

 * New preference to set size of dialog to create/edit text objects.

 * Fixed crash during stylesheet add (bug #315). 

 * New shortcut "Ctrl+Return" to accept dialogs (useful in multi-line
   text input fields).

 * Enabled display of paper boundary during panning again.

 * Menu item for showing/hiding bookmarks.

 * Tooltips in "Properties" dock.
 
 * "Edit" button in stylesheets dialog to edit stylesheet using
   external editor.

 * Fixed mouse shortcuts (Shift+Right and Shift+Ctrl+Right didn't work
   properly).

 * Moved path join/cap/fill rule to the path property menu.

 * Added opacity and gradient pattern to path property menu.

 * Current settings indicated in path property menu.

--------------------------------------------------------------------

Ipe 7.0.7 (2009-10-23) SVN 361

 * Removed -fno-exceptions and -fno-rtti from CXXFLAGS, to avoid
   surprises like bug #308.

 * Fixed broken interpretation of symbol/mark matrix (bug #310).

 * Fixed page numbering (bug #311).

 * Implemented "prefs.select_distance".

 * Fixed incorrect handling of line cap and line join (bug #312).

--------------------------------------------------------------------

Ipe 7.0.6 (2009-10-23) SVN 359

 * Worked around Qt styling bug in QComboBox (bug #308).  Using
   QSignalMapper now to map signals (no derived classes).

 * LayerBox popup moved into Lua code.

 * Fixed renaming of active layers.

 * Fixed selection of objects in locked layers.

 * Function to check symbolic attributes.

--------------------------------------------------------------------

Ipe 7.0.5 (2009-10-22) SVN 357

 * Created script "svgtoipe.py" to convert SVG files to Ipe format.
   It is available as a separate download.

 * Made Lua interface for attribute values more robust - interface now
   checks that attributes have correct type before passing them to
   Ipelib.  This fixes opacity names starting with digits. 
   (Bug #2880502)

 * #define to enable use of ancient pdftex versions ( < 130).
   Bug #309.

 * Show current view and page number in the properties window.

 * Environment variable IPEQTSTYLE sets this Qt style.  Saying, for
   instance, 'export IPEQTSTYLE="cleanlooks" ' would fix the Qt crash
   in some styles on some systems. (See bug #308.)


src/iperender/iperender.cpp
// --------------------------------------------------------------------
// iperender
// --------------------------------------------------------------------
/*

    This file is part of the extensible drawing editor Ipe.
    Copyright (C) 1993-2010  Otfried Cheong

    Ipe is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    As a special exception, you have permission to link Ipe with the
    CGAL library and distribute executables, as long as you follow the
    requirements of the Gnu General Public License in regard to all of
    the software in the executable aside from CGAL.

    Ipe is distributed in the hope that it will be useful, but WITHOUT
    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
    or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public
    License for more details.

    You should have received a copy of the GNU General Public License
    along with Ipe; if not, you can find it at
    "http://www.gnu.org/copyleft/gpl.html", or write to the Free
    Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

*/

#include "ipeutils.h"
#include "ipedoc.h"
#include "ipefonts.h"

#include "ipecairopainter.h"

#include <cstdio>
#include <cstdlib>

#include <cairo.h>
#include <cairo-svg.h>
#ifdef CAIRO_HAS_PDF_SURFACE
#include <cairo-pdf.h>
#endif
#ifdef CAIRO_HAS_PS_SURFACE
#include <cairo-ps.h>
#endif

enum TargetFormat { ESVG, EPNG, EPS, EPDF };

using ipe::Document;
using ipe::Page;

// --------------------------------------------------------------------

static void render(TargetFormat fm, const char *dst, const Document *doc,
		   const Page *page, int view, double zoom)
{
  ipe::Rect bbox = page->pageBBox(doc->cascade());
  int wid = int(bbox.width() * zoom + 1);
  int ht = int(bbox.height() * zoom + 1);

  unsigned long *data = 0;
  cairo_surface_t* surface = 0;

  if (fm == EPNG) {
    data = new unsigned long[wid * ht];
    for (unsigned long *p = data; p < data + wid * ht; )
      *p++ = 0xffffffff;
    surface = cairo_image_surface_create_for_data((uchar *) data,
						  CAIRO_FORMAT_ARGB32,
						  wid, ht, wid * 4);
  } else if (fm == ESVG) {
    surface = cairo_svg_surface_create(dst, wid, ht);
#ifdef CAIRO_HAS_PS_SURFACE
  } else if (fm == EPS) {
    surface = cairo_ps_surface_create(dst, wid, ht);
#endif
#ifdef CAIRO_HAS_PDF_SURFACE
  } else if (fm == EPDF) {
    surface = cairo_pdf_surface_create(dst, wid, ht);
#endif
  }

  cairo_t *cc = cairo_create(surface);
  cairo_scale(cc, zoom, -zoom);
  cairo_translate(cc, -bbox.topLeft().x, -bbox.topLeft().y);

  IpeAutoPtr<ipe::Fonts> fonts(ipe::Fonts::New(doc->fontPool()));
  ipe::CairoPainter painter(doc->cascade(), fonts.ptr(), cc, zoom, true);
  // painter.Transform(IpeLinear(zoom, 0, 0, -zoom));
  // painter.Translate(-bbox.TopLeft());
  painter.pushMatrix();
  for (int i = 0; i < page->count(); ++i) {
    if (page->objectVisible(view, i))
      page->object(i)->draw(painter);
  }
  painter.popMatrix();
  cairo_surface_flush(surface);
  cairo_show_page(cc);

  if (fm == EPNG)
    cairo_surface_write_to_png(surface, dst);

  cairo_destroy(cc);
  cairo_surface_destroy(surface);
  delete [] data;
}

// --------------------------------------------------------------------

static int renderPage(TargetFormat fm, const char *src, const char *dst,
		      int pageNum, int viewNum, double zoom)
{
  Document *doc = Document::loadWithErrorReport(src);

  if (!doc)
    return 1;

  if (pageNum < 1 || pageNum > doc->countPages()) {
    fprintf(stderr,
	    "The document contains %d pages, cannot convert page %d.\n",
	    doc->countPages(), pageNum);
    delete doc;
    return 1;
  }

  if (doc->runLatex()) {
    delete doc;
    return 1;
  }

  const Page *page = doc->page(pageNum - 1);
  render(fm, dst, doc, page, viewNum - 1, zoom);
  delete doc;
  return 0;
}

// --------------------------------------------------------------------

static void usage()
{
  fprintf(stderr,
	  "Usage: iperender [ -svg | -png ] "
	  "[ -page <page> ] [ -view <view> ] [ -resolution <dpi> ] "
	  "infile outfile\n"
	  "Iperender saves a single page of the Ipe document in some formats.\n"
	  " -page       : page to save (default 1).\n"
	  " -view       : view to save (default 1).\n"
	  " -resolution : resolution for png format (default 72.0 ppi).\n");
  exit(1);
}

int main(int argc, char *argv[])
{
  ipe::Platform::initLib(ipe::IPELIB_VERSION);

  // ensure at least three arguments (handles -help as well :-)
  if (argc < 4)
    usage();

  TargetFormat fm = EPNG;
  if (!strcmp(argv[1], "-png"))
    fm = EPNG;
#ifdef CAIRO_HAS_PS_SURFACE
  else if (!strcmp(argv[1], "-ps"))
    fm = EPS;
#endif
#ifdef CAIRO_HAS_PDF_SURFACE
  else if (!strcmp(argv[1], "-pdf"))
    fm = EPDF;
#endif
  else if (!strcmp(argv[1], "-svg"))
    fm = ESVG;
  else
    usage();

  int page = 1;
  int view = 1;
  double dpi = 72.0;
  int i = 2;

  if (!strcmp(argv[i], "-page")) {
    page = ipe::Lex(ipe::String(argv[i+1])).getInt();
    i += 2;
  }

  if (!strcmp(argv[i], "-view")) {
    view = ipe::Lex(ipe::String(argv[i+1])).getInt();
    i += 2;
  }

  if (!strcmp(argv[i], "-resolution")) {
    dpi = ipe::Lex(ipe::String(argv[i+1])).getDouble();
    i += 2;
  }

  // remaining arguments must be two filenames
  if (argc != i + 2)
    usage();

  const char *src = argv[i];
  const char *dst = argv[i+1];

  return renderPage(fm, src, dst, page, view, dpi / 72.0);
}

// --------------------------------------------------------------------
